### PR TITLE
Prepare for 2.x interface.

### LINKED ISSUE #204: Clarify null and undefined serialization binding behavior in docs
There are some nuances regarding binding `null` and `undefined` binding behavior in XElement — philosophically we want to leverage (and not augment) the browser's behavior.

Add detailed info here: https://github.com/Netflix/x-element/blob/main/doc/TEMPLATES.md

Maybe update the philosophy: "Please do not make me learn anything proprietary about the framework unless it is absolutely necessary!"

Consider:

```javascript
// uses element.textContent = binding
html`<div>${null}</div>`
// -> <div></div>
html`<div>${undefined}</div>`
// -> <div></div>


// uses element.innerHTML = binding (unsafe!)
html`<div>${unsafeHTML(null)}</div>`
// -> <div></div>
html`<div>${unsafeHTML(undefined)}</div>`
// -> <div>undefined</div>


// uses element.setAttribute('name', binding)
html`<div test-null="${null}"></div>`
// -> <div test-null="null"></div>
html`<div test-undefined="${undefined}"></div>`
// -> <div test-undefined="undefined"></div>


// uses element.setAttribute('name', binding)
html`<div test-null="${null ?? ''}"></div>`
// -> <div test-null></div>
html`<div test-undefined="${undefined ?? ''}"></div>`
// -> <div test-undefined></div>


// uses binding ? element.setAttribute('name', '') : element.removeAttribute('name') (boolean attribute behavior)
html`<div ?test-null="${null}"></div>`
// -> <div></div>
html`<div ?test-undefined="${undefined}"></div>`
// -> <div></div>


// magic helper (undefined)
/*
 if (binding === undefined) {
   element.removeAttribute('name')
 } else {
   element.setAttribute('name', binding)
}
*/
html`<div test-null="${ifDefined(null)}"></div>`
// -> <div test-null="null"></div>
html`<div test-undefined="${ifDefined(undefined)}"></div>`
// -> <div></div>


// magic helper (nullish)
/*
 if (binding === undefined || binding === null) {
   element.removeAttribute('name')
 } else {
   element.setAttribute('name', binding)
}
*/
html`<div test-null="${nullish(null)}"></div>`
// -> <div></div>
html`<div test-undefined="${nullish(undefined)}"></div>`
// -> <div></div>


// These cases throw an error
html`<div>${ifDefined(null)}</div>`
html`<div>${ifDefined(undefined)}</div>`
html`<div>${nullish(null)}</div>`
html`<div>${nullish(undefined)}</div>`
html`<div ?test-null="${ifDefined(null)}"></div>`
html`<div ?test-undefined="${ifDefined(undefined)}"></div>`
html`<div ?test-null="${nullish(null)}"></div>`
html`<div ?test-undefined="${nullish(undefined)}"></div>`

```

Perhaps the structure of the docs ought to be like "How do I ..."

* bind string content to the DOM
* bind real markup to the DOM
* bind string values to the DOM (attribute=value)
* bind boolean values to the DOM (presence of attribute name)
* only bind an attribute to the DOM when the value is not nullish (for important cases like `<a href="">`)

### PR DESCRIPTION
**Pull request by theengineear**
Changes:
* Adding `??attr` binding syntax.
* Changing `ifDefined` to delete attr on `null` as well as `undefined`.
* Deprecating `ifDefined` / `nullish` in favor of `??attr` binding.
* Deprecating `repeat` in favor of `map`.
* Updating documentation.

Closes #204.

### CODE REVIEW

File: test/test-template-engine.js [hunk 1]
```diff
@@ -729,6 +755,107 @@ describe('html updaters', () => {
...
+    it('can change from map to null content', () => run(toMap, toNullContent));
+    it('can change from map to text content', () => run(toMap, toTextContent));
+    it('can change from map to array content', () => run(toMap, toArrayContent));
+    it('can change from map to unsafe html', () => run(toMap, toUnsafeHTML));
+
+    it('can change from unsafeHtml to undefined content', () => run(toUnsafeHTML, toUndefinedContent));
+    it('can change from unsafeHtml to null content', () => run(toUnsafeHTML, toNullContent));
+    it('can change from unsafeHtml to text content', () => run(toUnsafeHTML, toTextContent));
+    it('can change from unsafeHtml to array content', () => run(toUnsafeHTML, toArrayContent));
+    it('can change from unsafeHtml to map', () => run(toUnsafeHTML, toMap));
```
Comment (theengineear) [inline test/test-template-engine.js:857]: Took me longer than I‘d like to admit to figure out how to write _expressive_ tests for this. I settled on this format, which feels pretty readable to me. Pretty happy with how this turned out – I ended up catching / fixing quite a few edge-case bugs. For example, previously if you bound `unsafeHTML(/* stuff */)` to some content and then bound `null` to that same content — the prior html didn’t actually get cleared for [reasons].

This should (at least minimally) test every possible content transition that we have today.

File: x-element.js [hunk 2]
```diff
@@ -1710,39 +1698,146 @@ class TemplateEngine {
...
+      case TemplateEngine.#ifDefined:
+        throw new Error(`The ifDefined update must be used on ${TemplateEngine.#getTypeText('attribute')}, not on ${TemplateEngine.#getTypeText(type)}.`);
+      case TemplateEngine.#nullish:
+        throw new Error(`The nullish update must be used on ${TemplateEngine.#getTypeText('attribute')}, not on ${TemplateEngine.#getTypeText(type)}.`);
+      case TemplateEngine.#repeat:
+        throw new Error(`The repeat update must be used on ${TemplateEngine.#getTypeText('content')}, not on ${TemplateEngine.#getTypeText(type)}.`);
+    }
+  }
+
+  static #resetIfRequired(node, startNode, value, lastValue, context, lastContext) {
```
Comment (theengineear) [inline x-element.js:1832]: In case you’re curious — this function fixes a _bunch_ of nuanced edge cases surrounding the transition from binding one kind of content value to another.

File: x-element.js [hunk 3]
```diff
@@ -1042,9 +1087,9 @@ class TemplateEngine {
    * @returns {any}
    */
   static html(strings, ...values) {
-    const symbol = Symbol();
-    const result = TemplateEngine.#createResult('html', strings, values);
-    TemplateEngine.#resultMap.set(symbol, result);
+    const symbol = Object.create(null);
```
Comment (theengineear) [inline x-element.js:1090]: Turns out… (1) FF still doesn’t support Symbols as WeakMap keys and (2) it’s more performant to call `Object.create(null)` for this purpose — theperfdontlie.

File: x-element.js [hunk 4]
```diff
@@ -1643,106 +1466,422 @@ class TemplateEngine {
...
-      if (Array.isArray(value)) {
-        const state = TemplateEngine.#setIfMissing(TemplateEngine.#stateMap, startNode, () => ({}));
-        TemplateEngine.#mapInner(state, node, startNode, identify, callback, value, name);
-      } else {
-        throw new Error(`Unexpected ${name} value "${value}".`);
+  static #map(node, startNode, value, identify, callback) {
+    TemplateEngine.#mapInner(node, startNode, identify, callback, value, 'map');
+  }
+
+  // Deprecated. Will remove in future release.
```
Comment (theengineear) [inline x-element.js:1473]: The git diff is pretty ugly in here… but… I’ve purposefully set all the deprecated stuff up so that it’ll be a breeze to remove in a follow-on PR once we get to `2.x`.

File: x-element.js [hunk 5]
```diff
@@ -1012,25 +1012,70 @@ export default class XElement extends HTMLElement {
...
+  // Special markers added to markup enabling discovery post-instantiation.
+  static #ATTRIBUTE_PREFIX = 'x-element-attribute';
+  static #BOOLEAN_PREFIX = 'x-element-boolean';
+  static #DEFINED_PREFIX = 'x-element-defined';
+  static #PROPERTY_PREFIX = 'x-element-property';
+  static #CONTENT_PREFIX = 'x-element-content';
+
+  // Patterns to find special edges in original html strings.
+  static #OPEN_REGEX = /<[a-z][a-z0-9-]*(?=\s)/g;
+  static #STEP_REGEX = /(?:\s+[a-z][a-z0-9-]*(?=[\s>])|\s+[a-z][a-zA-Z0-9-]*="[^"]*")+/y;
```
Comment (theengineear) [inline x-element.js:1024]: Small perf win here. Rather than match multiple times, we can let the single regex match more greedily by wrapping the whole thing in a non-capturing group and adding a `+` to the end.

File: x-element.js [hunk 6]
```diff
@@ -1012,25 +1012,70 @@ export default class XElement extends HTMLElement {
...
+  static #ATTRIBUTE_OR_PROPERTY_REGEX = /\s+(?:(?<questions>\?{0,2})?(?<attribute>([a-z][a-zA-Z0-9-]*))|\.(?<property>[a-z][a-zA-Z0-9_]*))="$/y;
```
Comment (theengineear) [inline x-element.js:1025]: Another small perf / readability win — it’s at least as performant to match _either_ and attribute _or_ a property in one pattern and with newly-baselined named capture groups, the code that leverages this pattern ends up being more readable.

File: x-element.js [hunk 7]
```diff
@@ -1012,25 +1012,70 @@ export default class XElement extends HTMLElement {
...
+  static #nodeToArrayState = new WeakMap();
+
+  // Mapping of tagged template function “strings” to caches computations.
+  static #stringsToAnalysis = new WeakMap();
+
+  // Mapping of opaque references to internal result objects.
+  static #symbolToResult = new WeakMap();
+
+  // Mapping of opaque references to internal update objects.
+  static #symbolToUpdate = new WeakMap();
```
Comment (theengineear) [inline x-element.js:1044]: I improved the naming of all the maps in this file. They’re just more explicit now.

File: x-element.js [hunk 8]
```diff
@@ -1643,106 +1466,422 @@ class TemplateEngine {
...
+            TemplateEngine.#clearObject(state);
+          }
+          const previousSibling = node.previousSibling;
+          if (previousSibling === startNode) {
+            // The `?? ''` is a shortcut for creating a text node and then
+            //  setting its textContent. It’s exactly equivalent to the
+            //  following code, but faster.
+            // const textNode = document.createTextNode('');
+            // textNode.textContent = value;
+            const textNode = document.createTextNode(value ?? '');
```
Comment (theengineear) [inline x-element.js:1827]: Had to walk back this fork removal. It’s just better performance to do this in one step. I left a fat comment about it though 😅

File: x-element.js [hunk 9]
```diff
@@ -1752,9 +1891,20 @@ class TemplateEngine {
...
+    const node = document.createComment('');
+    referenceNode.parentNode.insertBefore(startNode, referenceNode);
+    referenceNode.parentNode.insertBefore(node, referenceNode);
+    return { startNode, node };
+  }
+
+  static #insertAllBefore(parentNode, referenceNode, nodes) {
+    // Make a copy of the array, else the live NodeList will be mutated as you
+    //  iterate — which would cause us to miss nodes.
+    // Note that passing “null” as the reference node appends nodes to the end.
```
Comment (theengineear) [inline x-element.js:1905]: I also tried to leave commentary wherever I noticed patterns that might be non-obvious to readers — in this case, certain browser APIs return “live” collections which cannot be iterated over if you plan to do any mutating.

File: demo/lit-html/index.html [hunk 10]
```diff
@@ -2,6 +2,13 @@
 <html>
   <head>
     <meta charset="UTF-8">
+    <script type="importmap">
```
Comment (theengineear) [inline demo/lit-html/index.html:5]: I went ahead on modernized our demos to use import maps and bare-module specifiers for our dependencies.

File: eslint.config.js [hunk 11]
```diff
@@ -23,11 +19,6 @@ export default [
       'jsdoc/require-returns-description': 'off',
     },
   },
-  {
-    ...NetflixCommon,
-    files: ['demo/react/**/*.js'],
-    languageOptions: { globals: { ...globals.browser, React: 'readonly', ReactDOM: 'readonly' } },
```
Comment (theengineear) [inline eslint.config.js:29]: By importing react as an esm module — we can declare module-level variables and ditch this carve-out in `eslint` to relax rules for the react demo.

File: x-element.js [hunk 12]
```diff
@@ -1012,25 +1012,70 @@ export default class XElement extends HTMLElement {
...
+  static #symbolToResult = new WeakMap();
+
+  // Mapping of opaque references to internal update objects.
+  static #symbolToUpdate = new WeakMap(); 
+
+  /**
+   * Default template engine interface — what you get inside “template”.
+   * @type {{[key: string]: Function}}
+   */
+  static interface = Object.freeze({
```
Comment (theengineear) [inline x-element.js:1050]: The template engine `interface` keeps getting cleaner. Here’s where it’ll be at for `2.x`:

```js
{
    // Long-term interface.
    render: TemplateEngine.render,
    html: TemplateEngine.html,
    svg: TemplateEngine.svg,
    map: TemplateEngine.map,
    unsafe: TemplateEngine.unsafe,

    // Tentative interface.
    live: TemplateEngine.live,
}
```

File: doc/TEMPLATES.md [hunk 13]
```diff
@@ -20,56 +23,335 @@ static template(html, { map }) {
...
+**A note on non-primitive data:**
+
+Because DOM manipulation is *slow* — template engines do their best to avoid it
+(if a value hasn’t changed, DOM manipulation is skipped). To know if a
+non-primitive property has changed, a change-by-reference check is performed
+(versus a change-by-value check). Therefore, treat non-primitives as if they’re
+immutable. Do this: `element.property = { ...element.property, foo: 'bar' }`,
+and don’t do this: `element.property.foo = 'bar'`.
+
+## Value binding (in detail)
```
Comment (theengineear) [inline doc/TEMPLATES.md:96]: As discussed, I tried to clearly document all the different binding-value pairs that were sensible. We may need to shift some of this stuff around in the future, but I think this starting to be a pretty complete reference.

**Comment by theengineear** [comment]
@klebba — This PR kinda sucked me in… I’m going to get it merged so we can discuss and make smaller PRs against `main`. All-in-all, I ended up making a bunch of small-but-important fixes, re-vamping our interface, and re-tuning for performance.

I tried to document the high-level summary in `CHANGELOG.md`.
